### ISSUE TITLE
How doe the proposed architecture relate to the concept of an Auth proxy?

### ISSUE THREAD

**Issue by RubenVerborgh** [RubenVerborgh/solid-server-architecture#11]
A suggestion by @kjetilk was to design the architecture such that blocks of it can be used independently. One concrete idea was that Auth could be a layer on the HTTP level that can be transparently inserted in between. This would allow us to insert an Auth layer transparently, or even having it run on a different server.

In this issue, let's discuss about the relation of the proposed architecture in this repository to such an idea.

**Comment by RubenVerborgh** [comment]
### Background

Let's first detail the concept of an auth server. It would be an HTTP server, so an independent module that takes an HTTP request as input and returns an HTTP response as output. Internally, it would perform authorization and (basic) authentication, and act as a proxy server that then forwards authenticated requests. (Side-note: the current NSS can be configured to act as such as server, but of course then is also a full NSS server in addition to being a proxy.)

### Overview

In this reply, I will tackle two points:
1. Why I opted to _not_ design auth as an HTTP handler (= HTTP request in, response out).
2. How the concept of an auth proxy still can be realized with the proposed architecture.

However, I want to stress that a design where auth _is_ a proxy, is indeed a viable architecture and should be considered. I would very much like to have such an architectural diagram, so that we can meaningfully discuss trade-offs. Nonetheless, below I will sketch what I perceive to be the drawbacks.

### 1. Why the proposed architecture does not have Auth as an HTTP module

#### Problems with an HTTP handler-centric architecture in general

A major cause of unmaintainability within node-solid-server is the fact that _everything_ is written as an HTTP handler. So every single piece of functionality is written as a function that takes an HttpRequest and HttpResponse as parameters, and are chained together in a waterfall-style calling tree. This is an (unintended) extreme design choice, and I will zoom in on the resulting problems.

A first issue is that every component knows / needs to know about HTTP. This also means they can only be tested as an HTTP component, so typical tests more closely resemble integration tests than unit tests. While this can be advocated for functionality that is close to HTTP (such as auth), it is definitely detrimental for business logic.

A second issue is that all data passing necessarily happens through the HttpRequest (as this is the only input parameter shared across objects), which turn this HttpRequest into a God (data) object. All state (such as auth) is attached to the HttpRequest object, and there is no enforceable way of limiting read or write access to this object. So every handler _can_ do anything with an HttpRequest object, and eventually they inevitably also _will_, tying the functionality of all handlers very closely to one another. This again makes unit testing and guaranteeing correct behavior very difficult.

A third issue is that functionality is often repeated in multiple places. Especially if HttpRequest is _not_ used as a God object, several handlers will need to extract the same information again. This is only a minor problem from a performance perspective, but a much more important problem from the correctness perspective, as components' behavior for the same tasks might eventually start drifting apart, and bugs fixed in one place can remain in another. Plus we'd have to test all those things for all possible variations time and time again. (Bugs like this have bitten us really badly in node-solid-server.)

#### Problems with auth as a handler

Now we should be careful to not make a straw man argument, given that @kjetilk is not calling to make everything a handler, only auth. (Or are there more?)

However, I also want to avoid a slippery slope, where more and more components would become handlers, because then the above list of problems will start applying. So an important question is where we draw the line. If auth becomes a handler, which other components become handlers? (That is actually a genuine question, so feel free to reply.)

The concrete problems I see with auth as a proxy is that some tasks will be carried out multiple times. Again, not problematic necessarily from a performance perspective, but more from a consistency/correctness/security perspective: do all handlers do the same checks in the same way? One very concrete example here is the parsing of the target: what resource does the IRI point to? Auth would need to do it, and then LDP would need to do it again. In case you think the answer is trivial, let me remind you that we actually have [an entire class dedicated to this in NSS](https://github.com/solid/node-solid-server/blob/51cd35c045eb7dfc85e369cf45258df9a4bafb08/lib/resource-mapper.js), which took me a week to draft the initial [specs](https://github.com/solid/node-solid-server/blob/51cd35c045eb7dfc85e369cf45258df9a4bafb08/test/unit/resource-mapper-test.js) for, and which have since been edited by 5 people. This is a conceptually very simple question with a complex answer, and the auth proxy would need to do it exactly the same way like the layers below it. Did I mention that there are [security concerns](https://github.com/RubenVerborgh/solid-server-ts/blob/a93717ec02389ba10c79975fc2bbdbe8d18336f0/src/http/TargetExtractor.ts#L33-L43) with it too? And remember, parsing the target is just the easy example here of a shared task.

Now an easy way to address the above is to turn the tasks of parsing targets into a separate (non-handler) component that both handlers reuse. Well, sure enough, this is what I have in the architecture as [`TargetExtractor`](https://rubenverborgh.github.io/solid-server-architecture/solid-architecture-v1-0-0.pdf). So `Auth` and `LDP` could indeed just both use it as a shared component. But once you go down this route, what starts happening is that the `Auth` proxy just becomes a very simple wrapper around existing blocks—which brings me seamlessly to my next point.

### 2. How to realize an Auth HTTP module with the proposed architecture

There is definitely value in having an Auth module as an HTTP handler, not in the least to put it in front of legacy or existing systems. But this is actually quite easy to realize with the proposed architecture. So easy that I've written it below (draft):

```javascript
class AuthProxy {
  constructor(options) {
    Object.assign(this, options);
  }

  async handle(HttpRequest req, HttpResponse res) {
     const credentials = await this.credentialsExtractor.extractCredentials(req);
     const target = await this.targetExtractor.extractTarget(req);
     // Assuming just read here; if not, extract permissions from request with existing code
     if (this.authorizer.hasPermissions(credentials, target, PermissionSet.READ_ONLY))
       this.proxiedServer.handle(req, res);
     else
       res.end(credentials ? 403 : 401);
  }
}
```

An alternative implementation is to implement `ResourceStore` with an HTTP/LDP backend. More work, but can just be plugged in.

### Conclusion

Having Auth as an independent HTTP server is useful. If deployed independently, it comes with the overhead of the extra HTTP parsing, but that is probably negligible compared to the benefits it brings with regard to legacy systems.

However, the availability of Auth as an independent HTTP server does _not_ mean that we also should have it as such in the internal architecture of the server. Doing so would come with consistency challenges. The current design that has Auth as a non-HTTP module can very easily be turned into an auth proxy, as exemplified above; the reverse is much harder to achieve.

For me, having clean and specific interfaces between the core components (as opposed to a generic HttpRequest object) is more important because it brings increased code testability and maintainability, at a small cost of having a bit of extra code for an actual Auth HTTP proxy.

**TL;DR: Auth HTTP proxy: yes; necessarily also in the internal architecture: no.**

**Comment by kjetilk** [comment]
> A second issue is that all data passing necessarily happens through the HttpRequest (as this is the only input parameter shared across objects), which turn this HttpRequest into a God (data) object. All state (such as auth) is attached to the HttpRequest object, and there is no enforceable way of limiting read or write access to this object. So every handler _can_ do anything with an HttpRequest object, and eventually they inevitably also _will_, tying the functionality of all handlers very closely to one another. This again makes unit testing and guaranteeing correct behavior very difficult.

Right, but now you are tying yourself to what is available in Node/Express/NSS or something. From what I can tell, the Express HTTP handler is (intended to be?) a finite state machine, where indeed HttpRequest plays the role of a God object. That's their choice, but the argument cannot be generalized, surely, there has to exist frameworks where this is not the case, also frameworks where FSM is used without God objects?

Basically, I'm thinking there are three states that could come out of the proxy: Authorized, Unauthorized and Deferred, in the latter case, the authorization could not be determined because the proxy would not parse the body, and it is also the only case where the next layer would need to take precautions.


> 
> A third issue is that functionality is often repeated in multiple places. Especially if HttpRequest is _not_ used as a God object, several handlers will need to extract the same information again. This is only a minor problem from a performance perspective, but a much more important problem from the correctness perspective, as components' behavior for the same tasks might eventually start drifting apart, and bugs fixed in one place can remain in another. Plus we'd have to test all those things for all possible variations time and time again. (Bugs like this have bitten us really badly in node-solid-server.)

Again, this sounds like an argument against NSS that shouldn't have validity beyond that. The needed behaviour should be possible to encapsulate in unit-testable components that can be reused in different places as needed. I simply can't see any fundamental reason why this shouldn't work.

> #### Problems with auth as a handler
> 
> Now we should be careful to not make a straw man argument, given that @kjetilk is not calling to make everything a handler, only auth. (Or are there more?)

Well, I'm not calling for anything to be a handler, as Node, Express, etc are not my choices. I'm calling for a proxy. It could reasonably have been implemented in C for all I know. The complexity sits in WebID and authentication methods, the WAC could be implemented easily with serd.

> 
> The concrete problems I see with auth as a proxy is that some tasks will be carried out multiple times. Again, not problematic necessarily from a performance perspective, but more from a consistency/correctness/security perspective: do all handlers do the same checks in the same way? One very concrete example here is the parsing of the target: what resource does the IRI point to? Auth would need to do it, and then LDP would need to do it again. 

I'm really not following you here. They are really very different things, there's nothing like resource-mapper involved in such a proxy, it would only modify the authority component of the URL, nothing else. Auth simply asks the LDP layer to give it the ACL resources, which will be managed by a cache. Access is granted based on the URI, not the representation, so the representation of non-ACL resources will not need to be resolved. Or is it something else you are referring to?

Actually, the only duplication that is apparent is the need for authorization to happen for where the body must be parsed, i.e. `PATCH`. There is another optimization problem though, namely that the proxy would not know what resources exist before performing the WAC, but that too can be fixed.

What I sense here is that the limitations are closely tied to the choice of Node and HTTP handlers as implemented there. There is nothing fundamental about these problems, AFAICS, in fact, it seems like these are problems _introduced_ with the chosen frameworks, they would not have been a problem with a more generic Finite State Machine, AFAICS, even though that would have been similar in appearance.

I shall admit that I have been out of professional development for so long that I don't know the current state of the art, and so, I would need to spend quite a lot of time experimenting before I could come up with a feasible alternative. Moreover, Node is what we have, and if these problems are not easily overcome, then certainly, it would be infeasible to go that way.


> ### 2. How to realize an Auth HTTP module with the proposed architecture
> 
> There is definitely value in having an Auth module as an HTTP handler, not in the least to put it in front of legacy or existing systems. But this is actually quite easy to realize with the proposed architecture. So easy that I've written it below (draft):
> 
> ```js
> class AuthProxy {
>   constructor(options) {
>     Object.assign(this, options);
>   }
> 
>   async handle(HttpRequest req, HttpResponse res) {
>      const credentials = await this.credentialsExtractor.extractCredentials(req);
>      const target = await this.targetExtractor.extractTarget(req);
>      // Assuming just read here; if not, extract permissions from request with existing code
>      if (this.authorizer.hasPermissions(credentials, target, PermissionSet.READ_ONLY))
>        this.proxiedServer.handle(req, res);
>      else
>        res.end(credentials ? 403 : 401);
>   }
> }
> ```

Right, OK. This would bring along the entire server, which would make it a lot more code to audit if you want to run just a simple proxy, but that code certainly isn't bad.

At the end of the day, this is really about adoption, what would others be comfortable about using and adapting. I can't predict that, but I myself prefer relatively small components.

**Comment by RubenVerborgh** [comment]
I'm afraid we'll need to rewind because I neglected to make clear that my definition of “HTTP handler” is Node-independent.

> Right, but now you are tying yourself to what is available in Node/Express/NSS or something.

Definitely not: I intended “HTTP handler” generically as “a function that takes an HttpRequest and HttpResponse as parameters”. So an Express handler fits that definition, but any HTTP server or HTTP proxy also fit that definition. So an Auth proxy, even if a full-fledged HTTP server implemented in C, fits that definition.

> Basically, I'm thinking there are three states that could come out of the proxy: Authorized, Unauthorized and Deferred, in the latter case, the authorization could not be determined because the proxy would not parse the body, and it is also the only case where the next layer would need to take precautions.

I would actually suggest that an Auth proxy returns these bits of information (modeled as HTTP headers for the sake of example):
- `Agent-WebID: https://ruben.verborgh.org/profile/#me`
- `Validated-Permissions: Read, Write`

Just to be very explicit about exactly what has happened.


> > A third issue is that functionality is often repeated in multiple places.
>
> Again, this sounds like an argument against NSS that shouldn't have validity beyond that. The needed behaviour should be possible to encapsulate in unit-testable components that can be reused in different places as needed.

Not a NSS argument; a general argument for having components with an HTTP handler interface (as per the definition above) in the architecture.
Agreed on unit-testable components—but that is exactly the argument that I am making, and also my proposal in the second section. Because such unit-testable units will be pieces of functionality that are then reused by other layers as well, so these things will happen multiple times. Hence why am precisely arguing to not use the HTTP-based abstraction _inside_ of an architecture.

> I simply can't see any fundamental reason why this shouldn't work.

There is none; but maybe I should sharpen my points to make sure that we are discussing about the right things:
1. I think an Auth proxy as an auth module is a good idea, and that idea will work.
2. I think such an Auth proxy will consist of multiple unit-testable units, and those units will be reused by other components as well.
3. Because of that reuse, I suggest that _inside_ of a _complete_ Solid server implementation, I would _not_ put Auth proxy as an HTTP layer before LDP. (And this is my main point.) Doing that leads to double work (with potential for inconsistencies) and the usage of an HttpRequest God object. Rather, the internal interfaces should be more specific, and components should only see what they need, rather than the entire HTTP request.

> Well, I'm not calling for anything to be a handler, as Node, Express, etc are not my choices. I'm calling for a proxy.

These are conceptually the same thing for the purpose of my argumentation—does that make sense? If not, where is the fundamental difference, given that both are functions of the same arguments (Request/Response)?

> there's nothing like resource-mapper involved in such a proxy, it would only modify the authority component of the URL, nothing else. Auth simply asks the LDP layer to give it the ACL resources, which will be managed by a cache.

But that is not how the independent Auth HTTP Proxy (let's say written in C) would work, right? It would make decisions of its own, I presume?

> Access is granted based on the URI, not the representation

Agreed.

> so the representation of non-ACL resources will not need to be resolved.

Not the representation, but the ACL lookup will need to happen, and that requires mapping knowledge.

Let's go through a concrete example: a request for `/a/b/../x%20z?ghjd=dgdsg` comes in with a hostname of `bla.example.org`. Then the proxy will need to find the ACL for `bla.example.org/a/x z`, right? And after that, the LDP server will need to serve `bla.example.org/a/x z`?
Well, then we have just done the mapping twice—and better hope it was done consistently. I don't see another way around this.

> This would bring along the entire server

Nah, just the modules you need. They can be independent npm modules as far as I'm concerned (cfr. approach we are taking with Comunica).

> I can't predict that, but I myself prefer relatively small components.

We're not in disagreement if I take that literally 🙂
In the sense that, I think an Auth proxy is too big of a component to have _inside_ of a full Solid server implementation. I'd rather use its parts (as I do right now). I think it is a great component in its own right though.

**Comment by kjetilk** [comment]
> I'm afraid we'll need to rewind because I neglected to make clear that my definition of “HTTP handler” is Node-independent.

Hmmmm, OK!


> I would actually suggest that an Auth proxy returns these bits of information (modeled as HTTP headers for the sake of example):
> 
>     * `Agent-WebID: https://ruben.verborgh.org/profile/#me`
> 
>     * `Validated-Permissions: Read, Write`

Yeah, I can go with that. It is an exercise possibly worth doing: To see what variables would need to be passed in each state transition.

> 
> Just to be very explicit about exactly what has happened.
> 
> > > A third issue is that functionality is often repeated in multiple places.
> > 
> > 
> > Again, this sounds like an argument against NSS that shouldn't have validity beyond that. The needed behaviour should be possible to encapsulate in unit-testable components that can be reused in different places as needed.
> 
> Not a NSS argument; a general argument for having components with an HTTP handler interface (as per the definition above) in the architecture.
> Agreed on unit-testable components—but that is exactly the argument that I am making, and also my proposal in the second section. Because such unit-testable units will be pieces of functionality that are then reused by other layers as well, so these things will happen multiple times. Hence why am precisely arguing to not use the HTTP-based abstraction _inside_ of an architecture.
 
> > I simply can't see any fundamental reason why this shouldn't work.
> 
> There is none; 

OK, good! :-)

but maybe I should sharpen my points to make sure that we are discussing about the right things:
>
> 
>     1. I think an Auth proxy as an auth module is a good idea, and that idea will work.
> 
>     2. I think such an Auth proxy will consist of multiple unit-testable units, and those units will be reused by other components as well.
> 
>     3. Because of that reuse, I suggest that _inside_ of a _complete_ Solid server implementation, I would _not_ put Auth proxy as an HTTP layer before LDP. (And this is my main point.) Doing that leads to double work (with potential for inconsistencies) and the usage of an HttpRequest God object. Rather, the internal interfaces should be more specific, and components should only see what they need, rather than the entire HTTP request.
> 
> 
> > Well, I'm not calling for anything to be a handler, as Node, Express, etc are not my choices. I'm calling for a proxy.
> 
> These are conceptually the same thing for the purpose of my argumentation—does that make sense? If not, where is the fundamental difference, given that both are functions of the same arguments (Request/Response)?

Ah, so, the fundamental difference is the slippery slope you were arguing: That you'd happily shovel any variable you might need onto the HttpRequest object because of convenience, making it a God object. In a proxy, you would not consider doing that for the pragmatic reason there is no HttpRequest object, you'd have to serialize the variables and then pass them in the request to the next layer, where they would be parsed. I think this would raise a red flag to every programmer this is only something you should do with great care and a thoughtful process behind it, and thus, it would eliminate the problem we saw in NSS. 

I do think that there are things that would be passed, indeed, like the WebID, and that it could be done in a JWT with a shared secret for example, but the gap there would ensure that people aren't doing stuff like that in the way that they would in a Handler. The human element makes these two things different beasts.

> 
> > there's nothing like resource-mapper involved in such a proxy, it would only modify the authority component of the URL, nothing else. Auth simply asks the LDP layer to give it the ACL resources, which will be managed by a cache.
> 
> But that is not how the independent Auth HTTP Proxy (let's say written in C) would work, right? It would make decisions of its own, I presume?

Some, it could, but see below.


> > Access is granted based on the URI, not the representation
> 
> Agreed.
> 
> > so the representation of non-ACL resources will not need to be resolved.
> 
> Not the representation, but the ACL lookup will need to happen, and that requires mapping knowledge.
> 
> Let's go through a concrete example: a request for `/a/b/../x%20z?ghjd=dgdsg` comes in with a hostname of `bla.example.org`. Then the proxy will need to find the ACL for `bla.example.org/a/x z`, right? And after that, the LDP server will need to serve `bla.example.org/a/x z`?
> Well, then we have just done the mapping twice—and better hope it was done consistently. I don't see another way around this.

What would happen is that the proxy would already have a cached copy of the ACL in its memory quad store, and so it just runs against that. :-) 

So, what you're really asking is what the cache maintenance system would do. If it needs to request an ACL it doesn't have for some reason, say `https://bla.example.org/a/.acl`, it would look at the authority component of the URL, e.g. `bla.example.org`, look that up, find that it needs to change that to `10.0.192.1:8080` and it then requests `https://10.0.192.1:8080/a/.acl` from the LDP server. The actual mapping to a concrete representation would not need to be the concern of the proxy.

In later iterations of the proxy, it could add sharding and stuff like that for horizontal scalability by making the authority component mapping more sophisticated. It could also terminate TLS at that point, etc. In the long run, private data complicates caches a lot, and so, I'm thinking how we could have "a HAProxy of Solid". Getting there could mean several layers though.


> > This would bring along the entire server
> 
> Nah, just the modules you need. They can be independent npm modules as far as I'm concerned (cfr. approach we are taking with Comunica).

Right, OK.


> 
> > I can't predict that, but I myself prefer relatively small components.
> 
> We're not in disagreement if I take that literally slightly_smiling_face
> In the sense that, I think an Auth proxy is too big of a component to have _inside_ of a full Solid server implementation. I'd rather use its parts (as I do right now). I think it is a great component in its own right though.

OK, so, I think we might not have a common understanding of what needs to be in that proxy, but we're getting there. :-)

**Comment by pmcb55** [comment]
Lots of great discussion in here, but I'd like to focus on just one point for now, Ruben's first big assertion:

> So every single piece of functionality is written as a function that takes an HttpRequest and HttpResponse as parameters, and are chained together in a waterfall-style calling tree. This is an (unintended) extreme design choice, and I will zoom in on the resulting problems.

Firstly, I'd actually advocate abstracting *above* even HTTP, i.e. each '_function takes a Request and Response as parameters_'. Secondly, they aren't only 'chained together in a waterfall-sytle', they are composed into recognized industry design patterns, i.e. Enterprise Integration Patterns (EIPs).

Apache Camel does exactly these two things, i.e. it has a Message class that abstracts a Request (or Response) into something that looks very like a HTTP request (i.e. is has Headers and a Body), but is actually even more abstract: https://livebook.manning.com/#!/book/camel-in-action-second-edition/chapter-1/147. Personally I think this is an important abstraction, as we need to support non-HTTP integrations too (e.g. for MQSeries, or WebSockets, or LDAP, etc.).

(By the way, there is one big, very clever performance optimization that can be applied (when modules are collocated) to prevent what might seem like the necessary marshaling/unmarshaling of a request at each 'hop' in an EIP workflow - Camel implements it using the Exchange).

But I do definitely agree with Ruben's description of the 'God' object, and again I'd actually abstract his description slightly further. I'd assert that it not only contains the entire incoming request, but that it also includes meta-data injected into it at every single 'hop' (potentially) as it propagates through it's entire processing workflow. So yes, this may seem extremely 'dangerous' (as anyone can alter any part of the message (potentially, although there are ways to mitigate, see below)), but in my experience it's the only way to build truly flexible systems (as inevitably some microservice deep in the system will _tomorrow_ need just one extra parameter that _yesterday_ we didn't think it would ever need, so we never passed it, and now lots of interfaces have to change to pass this one extra parameter all the way down to this lowly microservice).

(Re. mitigating the 'danger' of passing around an all-encompassing God object: we can applying EIP filters at any stage to strip out sensitive data (or meta-data), and/or we could sign data (or meta-data) to prevent altering, and/or encrypt data.)

So yes, it's a trade-off, most definitely - but my opinion is that we need to architect for maximum flexibility.

Beyond that, I think we all agree on the need for unit-testable code, and in fact I think I'm agreeing with *both* of you in that:
1) I think Auth can be developed as an independent module
2) but that it also does _**not**_ need to depend on HTTP (instead, it depends on EIP Messages)

As always, it boils down to finding the right level of abstraction - and personally, I think EIPs (and Apache Camel's implementation) have nailed it!

Unfortunately I can't find any good EIP implementations already written in JavaScript/Node, although people have tried (e.g. https://github.com/Workable/eip). But if we're implementing from scratch anyway, then getting our 'first principles' right is crucial, and I think EIP is the best guide we have today.

**Comment by RubenVerborgh** [comment]
Getting there!

> Ah, so, the fundamental difference is the slippery slope you were arguing: That you'd happily shovel any variable you might need onto the HttpRequest object because of convenience, making it a God object. In a proxy, you would not consider doing that for the pragmatic reason there is no HttpRequest object, you'd have to serialize the variables and then pass them in the request to the next layer, where they would be parsed. I think this would raise a red flag to every programmer this is only something you should do with great care

I find that a strong assumption 😉

Because conceptually, the variables will turn up as HTTP headers.

And if this is just for one layer, well fine, that won't hurt. Just like one extra property on `HttpRequest` also wouldn't hurt.
The (actual, not rhetorical) question is where we draw the line. Why Auth as a proxy, but not something else?

> What would happen is that the proxy would already have a cached copy of the ACL in its memory quad store, and so it just runs against that. :-)

Understood; the proxy is then a bit more complex than I initially had assumed. I thought it would just be a subset of the Server modules, but there will thus be some modules unique to the proxy.

> So, what you're really asking is what the cache maintenance system would do. If it needs to request an ACL it doesn't have for some reason, say `https://bla.example.org/a/.acl`, it would look at the authority component of the URL, e.g. `bla.example.org`, look that up, find that it needs to change that to `10.0.192.1:8080` and it then requests `https://10.0.192.1:8080/a/.acl` from the LDP server. The actual mapping to a concrete representation would not need to be the concern of the proxy.

Alas, the proxy would still need to know the normalized URL of the resource to be able to look it up in the proxy.

A request for `//a/b/../x%20z?ghjd=dgdsg` with hostname `bla.example.org` comes in, which URL(s) in the ACL should be respected?
- http://bla.example.org//a/b/../x%20z?ghjd=dgdsg
- https://bla.example.org//a/b/../x%20z?ghjd=dgdsg
- http://bla.example.org//a/b/../x%20z
- https://bla.example.org//a/b/../x%20z
- http://bla.example.org//a/x%20z?ghjd=dgdsg
- https://bla.example.org//a/x%20z?ghjd=dgdsg
- http://bla.example.org//a/x%20z
- https://bla.example.org//a/x%20z
- http://bla.example.org/a/b/../x%20z?ghjd=dgdsg
- https://bla.example.org/a/b/../x%20z?ghjd=dgdsg
- http://bla.example.org/a/b/../x%20z
- https://bla.example.org/a/b/../x%20z
- http://bla.example.org/a/x%20z?ghjd=dgdsg
- https://bla.example.org/a/x%20z?ghjd=dgdsg
- http://bla.example.org/a/x%20z
- https://bla.example.org/a/x%20z

Whatever the answer, better pray that proxy and server agree 😉

> OK, so, I think we might not have a common understanding of what needs to be in that proxy, but we're getting there. :-)

Indeed, but that's fine. There's room for different kinds of proxies for sure.

My main question for you, @kjetilk, is:
_**Inside** of the vNext server, how strongly do you prefer having Auth as an HTTP layer, instead of a modularized component with a more specific and restricted interface (cfr. my diagram)?_

**Comment by RubenVerborgh** [comment]
@pmcb55 

> a Message class that abstracts a Request (or Response) into something that looks very like a HTTP request (i.e. is has Headers and a Body), but is actually even more abstract

I'm not opposed, and @michielbdejong has suggested HTTP-independence in the past.

To this point, I want to note that only very few components in my proposed architecture are tied to HTTP (and happy to abstract them even further, although I'd suggest for v2.0 though).

> in my experience it's the only way to build truly flexible systems

Not disagreeing, but we are _not_ building a highly flexible system here. Solid is different from other architectures in the sense that the bulk of the intelligence happens at the client. The server is very dumb.

> it's a trade-off, most definitely - but my opinion is that we need to architect for maximum flexibility.

I am not saying there are no cases for server flexibility; I am saying that maintainability is much higher on my list of architectural drivers for vNext. The flexibility in my architecture stems from the fact that other HTTP handlers can reuse components; and reuse is improved if components do not depend on highly specific parameters being passed around.

In other words, I am in favor of a different trade-off. This is my opinion for the vNext version in JavaScript; I see different trade-offs for other contexts.

> As always, it boils down to finding the right level of abstraction - and personally, I think EIPs (and Apache Camel's implementation) have nailed it!

I'm hesitant about carrying EIP through an entire server, certainly for JavaScript vNext.

**Comment by kjetilk** [comment]
> Getting there!
> 
> > Ah, so, the fundamental difference is the slippery slope you were arguing: That you'd happily shovel any variable you might need onto the HttpRequest object because of convenience, making it a God object. In a proxy, you would not consider doing that for the pragmatic reason there is no HttpRequest object, you'd have to serialize the variables and then pass them in the request to the next layer, where they would be parsed. I think this would raise a red flag to every programmer this is only something you should do with great care
> 
> I find that a strong assumption wink
> 
> Because conceptually, the variables will turn up as HTTP headers.

They would...? I was going to add them to the JWT :-)


> 
> And if this is just for one layer, well fine, that won't hurt. Just like one extra property on `HttpRequest` also wouldn't hurt.

I see your point, but whatever goes in there would be subject for the Solid spec, as my goal is interoperability, so that you can take any Auth* proxy and set it in front of whatever you have in the back.


> Understood; the proxy is then a bit more complex than I initially had assumed. I thought it would just be a subset of the Server modules, but there will thus be some modules unique to the proxy.

Well, you can of course just have the ACL cache in the server too. I think the ACL cache is absolutely crucial however you do it, we're running this algo for every request, and it has to run in main memory anyway.
 

> Whatever the answer, better pray that proxy and server agree wink

Heh, no prayers, please! :-) 

[RFC 3986 Section 6](https://tools.ietf.org/html/rfc3986#section-6), sets some constraints on URI normalization. We probably need more, and this has to be specified, it knowledge should not be in any single piece of code anyway.
 
> > OK, so, I think we might not have a common understanding of what needs to be in that proxy, but we're getting there. :-)
> 
> Indeed, but that's fine. There's room for different kinds of proxies for sure.
> 
> My main question for you, @kjetilk, is:
> _**Inside** of the vNext server, how strongly do you prefer having Auth as an HTTP layer, instead of a modularized component with a more specific and restricted interface (cfr. my diagram)?_

So, the thing is that I would like to challenge your notion of "inside". I think we're just looking at it rather differently. To me, it is just a bunch of components that when taken together form a Solid server. Inside is for entirely different problems, resourceStore is inside.

I think very few would go "yeah, the Solid server is exactly what we want", they will go "oh, great, we should be on Solid, how can we build Solid on what we have?"

So, my motivating problems are stuff like "how can Solid be used to restrict access to a Jitsi meet", "how can Nextcloud be made Solid aware with minimal changes", "how can legacy data be semantically lifted and made available on Solid", "how can Solid be used to manage access to content we already have"

AFAICS, these are things that should be pretty straightforward to do, but it seems like the current architecture would create hoops that those who aren't looking to run the whole server would have to jump through, and it seems like there is a lot of early adopters that would find that problematic. But it may just be that I don't know our early adopters good enough.

So, if Auth* has a very restrictive interface, what would that have to say for unexpected reuse?

Anyway, to @pmcb55 's point, I would love to make things HTTP independent too. I saw that pretty clearly when I worked in the IoT space that we really need that urgently, but like @RubenVerborgh I'm a bit hesitant about EIP for vNext, but it stems mainly from ignorance of concrete solutions, so I'm very open to be convinced.

**Comment by RubenVerborgh** [comment]
We can / might want to draw a diagram that shows how an AuthProxy can be realized using existing components.

I think that the proxy approach is more of a packaging/deployment diagram matter than it is a class diagram matter; classes will likely be very similar.
